nearcache.go

```diff
@@ -0,0 +1,692 @@
+	atomic.AddInt64(&rs.stats.Hits, 1)
+}
+
+func (rs *nearCacheRecordStore) incrementMisses() {
+	atomic.AddInt64(&rs.stats.Misses, 1)
+}
+
+func (rs *nearCacheRecordStore) incrementExpirations() {
+	atomic.AddInt64(&rs.stats.Expirations, 1)
+}
+
+func (rs *nearCacheRecordStore) incrementOwnedEntryMemoryCost(cost int64) {
+	atomic.AddInt64(&rs.stats.OwnedEntryMemoryCost, cost)
+}
+
+func (rs *nearCacheRecordStore) decrementOwnedEntryMemoryCost(cost int64) {
+	atomic.AddInt64(&rs.stats.OwnedEntryMemoryCost, -cost)
+}
+
+func (rs *nearCacheRecordStore) incrementOwnedEntryCount() {
+	atomic.AddInt64(&rs.stats.OwnedEntryCount, 1)
+}
+
+func (rs *nearCacheRecordStore) decrementOwnedEntryCount() {
+	atomic.AddInt64(&rs.stats.OwnedEntryCount, -1)
+}
+
+func (rs *nearCacheRecordStore) incrementInvalidations() {
+	atomic.AddInt64(&rs.stats.Invalidations, 1)
+}
+
+func (rs *nearCacheRecordStore) incrementInvalidationRequests() {
+	atomic.AddInt64(&rs.stats.invalidationRequests, 1)
+}
+
+func (rs *nearCacheRecordStore) getRecord(key interface{}) (*nearCacheRecord, bool) {
+	rs.recordsMu.Lock()
+	value, ok := rs.records[key]
+	rs.recordsMu.Unlock()
+	return value, ok
+}
+
+func (rs *nearCacheRecordStore) recordExpired(rec *nearCacheRecord, nowMS int64) bool {
+	if !rs.canUpdateStats(rec) {
+		// A record can only be checked for expiry if its record state is READ_PERMITTED.
+		// We can't check reserved records for expiry.
+		return false
+	}
+	if rec.IsExpiredAtMS(nowMS) {
+		return true
+	}
+	return rec.IsIdleAtMS(rs.maxIdleMillis, nowMS)
+}
+
+func (rs *nearCacheRecordStore) canUpdateStats(rec *nearCacheRecord) bool {
+	return rec != nil && rec.ReservationID() == nearCacheRecordReadPermitted
+}
+
+func (rs *nearCacheRecordStore) TryReserveForUpdate(key interface{}, keyData serialization.Data, ups nearCacheUpdateSemantic) (int64, error) {
+	// checkAvailable()
+	// if there is no eviction configured we return if the Near Cache is full and it's a new key.
+	// we have to check the key, otherwise we might lose updates on existing keys.
+	/*
+	   if (evictionDisabled && evictionChecker.isEvictionRequired() && !containsRecordKey(key)) {
+	       return NOT_RESERVED;
+	   }
+	*/
+	rid := rs.nextReservationID()
+	var rec *nearCacheRecord
+	var err error
+	if ups == nearCacheUpdateSemanticWriteUpdate {
+		rec, err = rs.reserveForWriteUpdate(key, keyData, rid)
+	} else {
+		rec, err = rs.reserveForReadUpdate(key, keyData, rid)
+	}
+	if err != nil {
+		return 0, err
+	}
+	if rec == nil || rec.ReservationID() != rid {
+		return nearCacheRecordNotReserved, nil
+	}
+	return rid, nil
+}
+
+func (rs *nearCacheRecordStore) nextReservationID() int64 {
+	return atomic.AddInt64(&rs.reservationID, 1)
+}
+
+func (rs *nearCacheRecordStore) reserveForWriteUpdate(key interface{}, keyData serialization.Data, reservationID int64) (*nearCacheRecord, error) {
+	rs.recordsMu.Lock()
+	defer rs.recordsMu.Unlock()
+	rec, ok := rs.records[key]
```

Comment:
Is doing this better? 

```suggestion
	rs.recordsMu.Lock()
	rec, ok := rs.records[key]
	rs.recordsMu.Unlock()
```

Reply:
Like above the record is mutated in this method. I think it's better to do that in the critical section.